### DISCUSSION TITLE
General improvements to the CLI

### DISCUSSION THREAD

**Discussion by jonasraoni** [pkp/pkp-lib#7980] [Proposals]
**Describe the problem you would like to solve**
Based on the [Scout tool](https://github.com/henriqueramos/scout), there's a lot of possible improvements that can be done to the CLI tools.

- Automatic formatting and documentation of commands and arguments by using a 3rd-party package (Symfony CLI sounds like the best option, if it's too heavy, there's simpler options out there http://getopt-php.github.io/getopt-php).
- Having a generic CLI tool, packed in a `.phar` file, would allow administrators to possibly manage instances in a single place.
- Analyze ways to let the CLI tool discover operations available in the installation instead of coding features directly on it, this way it will be version agnostic (at least from now on...):
  - Discovery through the file system (CLI access the installation and looks for "tools" inside the default folders)
  - Create a `/tools` route and re-use the swagger file to support the CLI with operations, parameters and documentation. Looks doable, but will require a working URL (working installation or fake server).

**Who is asking for this feature?**
Developers, system administrators

**Comment by NateWr** [comment]
> let the CLI tool discover operations available in the installation 

My preference would be to find a way to make the CLI a wrapper for the API. We already have a comprehensive JSON description of the API, so we can automatically import argument descriptions and response formats from that.

If we can't use the OpenAPI doc itself (`swagger-source.json`), we might be able to use the JSON schemas of each entity directly to otherwise cut down on code duplication. One thing the API does that the CLI doesn't need to do is a lot of authorization checks, so if a CLI tool could use an entity repository programatically (ie - discover the CRUD commands), that might be a good alternative.
